lightning/ddl may panic when partial write and meet "needRescan" error

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

comment the code change part, and run the unit test in #43364 . The unit test case is that there an engine with 2 KV and regionSplitSize = 1, after partially writing the first KV the job continues to `ingest` function and then meet NotLeader error (other error will also trigger this problem)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

Because NotLeader error needs lightning to rescan the region, the worker should exit `executeJob`.

### 3. What did you see instead (Required)

We didn't check the stage of the job, and "partial write" logic let the job continue `writeToTiKV` in next round. But the region information is already cleared so triigger a nil-dereference

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x39747b6]

goroutine 472272 [running]:
[github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).writeToTiKV](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).writeToTiKV)(0xc011760c60, {0x5790be8, 0xc0b17e91d0}, 0xc0b2c3da40)
	/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/br/pkg/lightning/backend/local/region_job.go:185](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local/region_job.go:185) +0x156
[github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).executeJob](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).executeJob)(0xc011760c60, {0x5790be8, 0xc0b17e91d0}, 0xc0b2c3da40)
	/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1347](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1347) +0xb4
[github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).startWorker](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).startWorker)(0x438807e?, {0x5790be8, 0xc0b17e91d0}, 0xc01505aa80, 0xc0969d4d80?, 0xc06bb94718?)
	/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1261](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1261) +0x125
[github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doImport.func3()](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doImport.func3())
	/home/jenkins/agent/workspace/build-common/go/src/[github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1529](http://github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1529) +0x31
[golang.org/x/sync/errgroup.(*Group).Go.func1()](http://golang.org/x/sync/errgroup.(*Group).Go.func1())
	/go/pkg/mod/[golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75](http://golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75) +0x64
created by [golang.org/x/sync/errgroup.(*Group).Go](http://golang.org/x/sync/errgroup.(*Group).Go)
	/go/pkg/mod/[golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72](http://golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72) +0xa5
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/br/pkg/lightning/backend/local/local.go
+++ b/br/pkg/lightning/backend/local/local.go
@@ -1367,6 +1367,9 @@ func (local *Backend) executeJob(
 			job.lastRetryableErr = err
 			return nil
 		}
+		if job.stage == needRescan {
+			return nil
+		}
 
 		if job.writeResult == nil || job.writeResult.remainingStartKey == nil {
 			return nil

--- a/br/pkg/lightning/backend/local/local_test.go
+++ b/br/pkg/lightning/backend/local/local_test.go
@@ -1343,8 +1343,113 @@ func TestNotLeaderErrorNeedUpdatePeers(t *testing.T) {
 	// then meet NotLeader error, scanned new region (11,12,13)
 	// repeat above for 11,12,13
 	require.Equal(t, []uint64{1, 2, 3, 11, 12, 13}, apiInvokeRecorder["Write"])
-	// store 12 has a follower busy, so it will break the workflow for region (11, 12, 13)
 	require.Equal(t, []uint64{1, 2, 3, 1, 11, 12, 13, 11}, apiInvokeRecorder["MultiIngest"])
+}
+
+func TestPartialWriteIngestErrorWillPanic(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// let lightning meet any error that will call convertStageTo(needRescan)
+	apiInvokeRecorder := map[string][]uint64{}
+	notLeaderResp := &sst.IngestResponse{
+		Error: &errorpb.Error{
+			NotLeader: &errorpb.NotLeader{Leader: &metapb.Peer{StoreId: 11}},
+		}}
+
+	local := &Backend{
+		splitCli: initTestSplitClient3Replica([][]byte{{}, {'c'}}, nil),
+		importClientFactory: &mockImportClientFactory{
+			stores: []*metapb.Store{
+				{Id: 1}, {Id: 2}, {Id: 3},
+			},
+			createClientFn: func(store *metapb.Store) sst.ImportSSTClient {
+				importCli := newMockImportClient()
+				importCli.store = store
+				importCli.apiInvokeRecorder = apiInvokeRecorder
+				if store.Id == 1 {
+					importCli.retry = 1
+					importCli.resp = notLeaderResp
+				}
+				return importCli
+			},
+		},
+		logger:             log.L(),
+		writeLimiter:       noopStoreWriteLimiter{},
+		bufferPool:         membuf.NewPool(),
+		supportMultiIngest: true,
+		tikvCodec:          keyspace.CodecV1,
+	}
+
+	db, tmpPath := makePebbleDB(t, nil)
+	_, engineUUID := backend.MakeUUID("ww", 0)
+	engineCtx, cancel2 := context.WithCancel(context.Background())
+	f := &Engine{
+		db:           db,
+		UUID:         engineUUID,
+		sstDir:       tmpPath,
+		ctx:          engineCtx,
+		cancel:       cancel2,
+		sstMetasChan: make(chan metaOrFlush, 64),
+		keyAdapter:   noopKeyAdapter{},
+		logger:       log.L(),
+	}
+	err := f.db.Set([]byte("a"), []byte("a"), nil)
+	require.NoError(t, err)
+	err = f.db.Set([]byte("a2"), []byte("a2"), nil)
+	require.NoError(t, err)
+
+	jobCh := make(chan *regionJob, 10)
+
+	partialWriteJob := &regionJob{
+		keyRange: Range{start: []byte("a"), end: []byte("c")},
+		region: &split.RegionInfo{
+			Region: &metapb.Region{
+				Id: 1,
+				Peers: []*metapb.Peer{
+					{Id: 1, StoreId: 1}, {Id: 2, StoreId: 2}, {Id: 3, StoreId: 3},
+				},
+				StartKey: []byte("a"),
+				EndKey:   []byte("c"),
+			},
+			Leader: &metapb.Peer{Id: 1, StoreId: 1},
+		},
+		stage:  regionScanned,
+		engine: f,
+		// use small regionSplitSize to trigger partial write
+		regionSplitSize: 1,
+	}
+	var jobWg sync.WaitGroup
+	jobWg.Add(1)
+	jobCh <- partialWriteJob
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	jobOutCh := make(chan *regionJob)
+	go func() {
+		defer wg.Done()
+		for {
+			job := <-jobOutCh
+			if job.stage == regionScanned {
+				jobWg.Done()
+				return
+			}
+			require.Fail(t, "job stage %s is not expected", job.stage)
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := local.startWorker(ctx, jobCh, jobOutCh, &jobWg)
+		require.NoError(t, err)
+	}()
+
+	jobWg.Wait()
+	cancel()
+	wg.Wait()
+
+	require.Equal(t, []uint64{1, 2, 3}, apiInvokeRecorder["Write"])
+	require.Equal(t, []uint64{1}, apiInvokeRecorder["MultiIngest"])
 }
 
 // mockGetSizeProperties mocks that 50MB * 20 SST file.

